File: bgfx_study/app/src/main/cpp/src/samples/curbes.cpp
//
// Created by Administrator on 2020/8/2 0002.
//

#include "include/curbes.h"
#include "bgfx_utils.h"
#include "bx/timer.h"

struct PosColorVertex
{
    float m_x;
    float m_y;
    float m_z;
    uint32_t m_abgr;

    static void init()
    {
        ms_layout
                .begin()
                .add(bgfx::Attrib::Position, 3, bgfx::AttribType::Float)
                .add(bgfx::Attrib::Color0,   4, bgfx::AttribType::Uint8, true)
                .end();
    };

    static bgfx::VertexLayout ms_layout;
};

bgfx::VertexLayout PosColorVertex::ms_layout;

static PosColorVertex s_cubeVertices[] =
        {
                {-1.0f,  1.0f,  1.0f, 0xff000000 },
                { 1.0f,  1.0f,  1.0f, 0xff0000ff },
                {-1.0f, -1.0f,  1.0f, 0xff00ff00 },
                { 1.0f, -1.0f,  1.0f, 0xff00ffff },
                {-1.0f,  1.0f, -1.0f, 0xffff0000 },
                { 1.0f,  1.0f, -1.0f, 0xffff00ff },
                {-1.0f, -1.0f, -1.0f, 0xffffff00 },
                { 1.0f, -1.0f, -1.0f, 0xffffffff },
        };

static const uint16_t s_cubeTriList[] =
        {
                0, 1, 2, // 0
                1, 3, 2,
                4, 6, 5, // 2
                5, 6, 7,
                0, 2, 4, // 4
                4, 2, 6,
                1, 5, 3, // 6
                5, 7, 3,
                0, 4, 1, // 8
                4, 5, 1,
                2, 3, 6, // 10
                6, 3, 7,
        };

static const uint16_t s_cubeTriStrip[] =
        {
                0, 1, 2,
                3,
                7,
                1,
                5,
                0,
                4,
                2,
                6,
                7,
                4,
                5,
        };

static const uint16_t s_cubeLineList[] =
        {
                0, 1,
                0, 2,
                0, 4,
                1, 3,
                1, 5,
                2, 3,
                2, 6,
                3, 7,
                4, 5,
                4, 6,
                5, 7,
                6, 7,
        };

static const uint16_t s_cubeLineStrip[] =
        {
                0, 2, 3, 1, 5, 7, 6, 4,
                0, 2, 6, 4, 5, 7, 3, 1,
                0,
        };

static const uint16_t s_cubePoints[] =
        {
                0, 1, 2, 3, 4, 5, 6, 7
        };

static const char* s_ptNames[]
        {
                "Triangle List",
                "Triangle Strip",
                "Lines",
                "Line Strip",
                "Points",
        };
BX_STATIC_ASSERT(BX_COUNTOF(s_ptState) == BX_COUNTOF(s_ptNames) );

namespace heaven7_Bgfx_demo{

    void CurbesDemo::init(InitConfig *config) {
        m_debug  = BGFX_DEBUG_NONE;
        m_reset  = BGFX_RESET_VSYNC;
        BaseDemo::init(config);

        // Create vertex stream declaration.
        PosColorVertex::init();

        // Create static vertex buffer.
        m_vbh = bgfx::createVertexBuffer(
                // Static data can be passed with bgfx::makeRef
                bgfx::makeRef(s_cubeVertices, sizeof(s_cubeVertices) )
                , PosColorVertex::ms_layout
        );

        // Create static index buffer for triangle list rendering.
        m_ibh[0] = bgfx::createIndexBuffer(
                // Static data can be passed with bgfx::makeRef
                bgfx::makeRef(s_cubeTriList, sizeof(s_cubeTriList) )
        );

        // Create static index buffer for triangle strip rendering.
        m_ibh[1] = bgfx::createIndexBuffer(
                // Static data can be passed with bgfx::makeRef
                bgfx::makeRef(s_cubeTriStrip, sizeof(s_cubeTriStrip) )
        );

        // Create static index buffer for line list rendering.
        m_ibh[2] = bgfx::createIndexBuffer(
                // Static data can be passed with bgfx::makeRef
                bgfx::makeRef(s_cubeLineList, sizeof(s_cubeLineList) )
        );

        // Create static index buffer for line strip rendering.
        m_ibh[3] = bgfx::createIndexBuffer(
                // Static data can be passed with bgfx::makeRef
                bgfx::makeRef(s_cubeLineStrip, sizeof(s_cubeLineStrip) )
        );

        // Create static index buffer for point list rendering.
        m_ibh[4] = bgfx::createIndexBuffer(
                // Static data can be passed with bgfx::makeRef
                bgfx::makeRef(s_cubePoints, sizeof(s_cubePoints) )
        );

        // Create program from shaders.
        m_program = loadProgram("vs_cubes", "fs_cubes");

        m_timeOffset = bx::getHPCounter();
    }
    void CurbesDemo::draw() {
        int m_width = config.win_width;
        int m_height = config.win_height;
        float time = (float)( (bx::getHPCounter()-m_timeOffset)/double(bx::getHPFrequency() ) );

        const bx::Vec3 at  = { 0.0f, 0.0f,   0.0f };
        const bx::Vec3 eye = { 0.0f, 0.0f, -35.0f };

        // Set view and projection matrix for view 0.
        {
            float view[16];
            bx::mtxLookAt(view, eye, at);

            float proj[16];
            bx::mtxProj(proj, 60.0f, float(m_width)/float(m_height), 0.1f, 100.0f, bgfx::getCaps()->homogeneousDepth);
            bgfx::setViewTransform(0, view, proj);

            // Set view 0 default viewport.
            bgfx::setViewRect(0, 0, 0, uint16_t(m_width), uint16_t(m_height) );
        }

        // This dummy draw call is here to make sure that view 0 is cleared
        // if no other draw calls are submitted to view 0.
        bgfx::touch(0);

        bgfx::IndexBufferHandle ibh = m_ibh[m_pt];
        uint64_t state = 0
                         | (m_r ? BGFX_STATE_WRITE_R : 0)
                         | (m_g ? BGFX_STATE_WRITE_G : 0)
                         | (m_b ? BGFX_STATE_WRITE_B : 0)
                         | (m_a ? BGFX_STATE_WRITE_A : 0)
                         | BGFX_STATE_WRITE_Z
                         | BGFX_STATE_DEPTH_TEST_LESS
                         | BGFX_STATE_CULL_CW
                         | BGFX_STATE_MSAA
                         | s_ptState[m_pt]
        ;

        // Submit 11x11 cubes.
        for (uint32_t yy = 0; yy < 11; ++yy)
        {
            for (uint32_t xx = 0; xx < 11; ++xx)
            {
                float mtx[16];
                bx::mtxRotateXY(mtx, time + xx*0.21f, time + yy*0.37f);
                mtx[12] = -15.0f + float(xx)*3.0f;
                mtx[13] = -15.0f + float(yy)*3.0f;
                mtx[14] = 0.0f;

                // Set model matrix for rendering.
                bgfx::setTransform(mtx);

                // Set vertex and index buffer.
                bgfx::setVertexBuffer(0, m_vbh);
                bgfx::setIndexBuffer(ibh);

                // Set render states.
                bgfx::setState(state);

                // Submit primitive for rendering to view 0.
                bgfx::submit(0, m_program);
            }
        }

        // Advance to next frame. Rendering thread will be kicked to
        // process submitted rendering primitives.
        bgfx::frame();
    }
    void CurbesDemo::destroy() {
// Cleanup.
        for (uint32_t ii = 0; ii < BX_COUNTOF(m_ibh); ++ii)
        {
            bgfx::destroy(m_ibh[ii]);
        }

        bgfx::destroy(m_vbh);
        bgfx::destroy(m_program);

        BaseDemo::destroy();
    }

}

File: bgfx_study/app/src/main/assets/lua/curbes.lua
local bgfx = require("bgfx");
local bx = require("bx");
local mem = require("memory");
local lib64 = require("uint64");
local int64 = lib64.new;

local initializer = bgfx.getInit();
local reso = initializer.resolution;

--------- fields ----------
local m_timeOffset;
local m_pt = 1;

local m_r = true;
local m_g = true;
local m_b = true;
local m_a = true;

local m_vbh;
local m_ibh = {};
local m_program;

------------ bgfx  ----------------
local ms_layout = bgfx.newVertexLayout();
local s_cubeVertices = mem.newMemoryFFFUI( {-1.0,  1.0,  1.0, 0xff000000},
        { 1.0,  1.0,  1.0, 0xff0000ff },
        {-1.0, -1.0,  1.0, 0xff00ff00 },
        { 1.0, -1.0,  1.0, 0xff00ffff },
        {-1.0,  1.0, -1.0, 0xffff0000 },
        { 1.0,  1.0, -1.0, 0xffff00ff },
        {-1.0, -1.0, -1.0, 0xffffff00 },
        { 1.0, -1.0, -1.0, 0xffffffff });

local s_cubeTriList = mem.newMemory('w', {
    0, 1, 2, -- 0
1, 3, 2,
4, 6, 5, -- 2
5, 6, 7,
0, 2, 4, -- 4
4, 2, 6,
1, 5, 3, -- 6
5, 7, 3,
0, 4, 1, -- 8
4, 5, 1,
2, 3, 6, -- 10
6, 3, 7,
})

local s_cubeTriStrip = mem.newMemory('w', {
    0, 1, 2,
    3,
    7,
    1,
    5,
    0,
    4,
    2,
    6,
    7,
    4,
    5,
})
local s_cubeLineList = mem.newMemory('w', {
    0, 1,
    0, 2,
    0, 4,
    1, 3,
    1, 5,
    2, 3,
    2, 6,
    3, 7,
    4, 5,
    4, 6,
    5, 7,
    6, 7,
})
local s_cubeLineStrip = mem.newMemory("w", {
    0, 2, 3, 1, 5, 7, 6, 4,
    0, 2, 6, 4, 5, 7, 3, 1,
    0,
})
local s_cubePoints = mem.newMemory('w', {
    0, 1, 2, 3, 4, 5, 6, 7
})
local s_ptNames = {
    "Triangle List",
    "Triangle Strip",
    "Lines",
    "Line Strip",
    "Points",
};
---  UINT64_C(0),
--                BGFX_STATE_PT_TRISTRIP,
--                BGFX_STATE_PT_LINES,
--                BGFX_STATE_PT_LINESTRIP,
--                BGFX_STATE_PT_POINTS,
local s_ptState = {
    0,
    int64(0x0001000000000000),
    int64(0x0002000000000000),
    int64(0x0003000000000000),
    int64(0x0004000000000000),
};

local BGFX_STATE_WRITE_R = int64(0x0000000000000001);
local BGFX_STATE_WRITE_G = int64(0x0000000000000002);
local BGFX_STATE_WRITE_B = int64(0x0000000000000004);
local BGFX_STATE_WRITE_A = int64(0x0000000000000008);
local BGFX_STATE_WRITE_Z = int64(0x0000004000000000);
local BGFX_STATE_DEPTH_TEST_LESS = int64(0x0000000000000010);
local BGFX_STATE_CULL_CW =  int64(0x0000001000000000);
local BGFX_STATE_CULL_CCW = int64(0x0000002000000000);
local BGFX_STATE_MSAA = int64(0x0100000000000000);
---------------------------
---
local app;
local app_pre_init = function()
    print("app_pre_init");
    initializer.type = "Count";
    initializer.vendorId = 0;
    reso.reset = 0x00000080; --vsync
    print('end app_pre_init ----- ')
end
local app_init = function ()
    print("app_init");
    bgfx.setDebug(0); -- BGFX_DEBUG_NONE
    bgfx.setViewClear(0
    , 3          -- BGFX_CLEAR_COLOR|BGFX_CLEAR_DEPTH
    , 0x303030ff
    , 1.0
    , 0
    );
    -- init layout
    ms_layout
            .begin()
            .add("Position", 3, "Float")
            .add("Color0",   4, "Uint8", true)
            .ends();
    m_vbh = bgfx.createVertexBuffer(bgfx.makeRef(s_cubeVertices), ms_layout);
    m_ibh[1] = bgfx.createIndexBuffer(bgfx.makeRef(s_cubeTriList));
    m_ibh[2] = bgfx.createIndexBuffer(bgfx.makeRef(s_cubeTriStrip));
    m_ibh[3] = bgfx.createIndexBuffer(bgfx.makeRef(s_cubeLineList));
    m_ibh[4] = bgfx.createIndexBuffer(bgfx.makeRef(s_cubeLineStrip));
    m_ibh[5] = bgfx.createIndexBuffer(bgfx.makeRef(s_cubePoints));

    m_program = bgfx.loadProgram("vs_cubes", "fs_cubes");

    m_timeOffset = bx.getHPCounter();
    print("app_init done, m_timeOffset = ", m_timeOffset);
end

local logFirst = true;
--local time = 0;
local app_draw = function ()
    print("app_draw");
    local m_width = reso.width;
    local m_height = reso.height;

    --time = time + 1.0/60.0;
    local time = (bx.getHPCounter()-m_timeOffset)*1.0/bx.getHPFrequency();
    if(logFirst) then
        print("time: ", time)
    end
    --
    local at = bx.newVec3(0, 0, 0);
    local eye = bx.newVec3(0, 0, -35);

    --Set view and projection matrix for view 0.
    local view = mem.newMemory('f', 16);
    bx.mtxLookAt(view, eye, at);
    if(logFirst) then
        print("mtxLookAt: "..tostring(view));
    end

    local proj = mem.newMemory('f', 16);
    --bx::mtxProj(proj, 60.0f, float(m_width)/float(m_height), 0.1f, 100.0f, bgfx::getCaps()->homogeneousDepth);
    bx.mtxProj(proj, 60, m_width / m_height, 0.1, 100, bgfx.getCaps().homogeneousDepth);
    if(logFirst) then
        print("mtxProj: "..tostring(proj));
    end

    bgfx.setViewTransform(0, view, proj);
    --Set view 0 default viewport.
    bgfx.setViewRect(0, 0, 0, m_width, m_height);

    -- This dummy draw call is here to make sure that view 0 is cleared
    -- if no other draw calls are submitted to view 0.
    bgfx.touch(0);
    local ibh = m_ibh[m_pt];

    local state = int64(0) | (m_r and BGFX_STATE_WRITE_R or 0)
         | (m_g and BGFX_STATE_WRITE_G or 0)
         | (m_b and BGFX_STATE_WRITE_B or 0)
         | (m_a and BGFX_STATE_WRITE_A or 0)
         | BGFX_STATE_WRITE_Z
         | BGFX_STATE_DEPTH_TEST_LESS
         | BGFX_STATE_CULL_CW
         | BGFX_STATE_MSAA
         | s_ptState[m_pt]
    --print("state", state)
    if(logFirst) then
        print("state: "..tostring(state));
    end

    if(logFirst) then
        print("----- start 11*11 curbes -----\n")
    end
    for y = 1, 11, 1 do
        local yy = y - 1;
        for x = 1, 11, 1 do
            --print("y, x = ", y, x)
            local xx = x - 1;
            local mtx = mem.newMemory('f', 16);
            --print("pre mtxRotateXY: "..tostring(mtx));
            bx.mtxRotateXY(mtx, time + xx*0.21, time + yy*0.37);
            --
            mtx[12] = -15.0 + xx*3.0;
            mtx[13] = -15.0 + yy*3.0;
            mtx[14] = 0.0;
            if(logFirst) then
                print("(xx, yy) = (".. xx..", "..yy..")")
                print("mtxRotateXY: "..tostring(mtx));
            end

            -- Set model matrix for rendering.
            bgfx.setTransform(mtx);

            -- Set vertex and index buffer.
            bgfx.setVertexBuffer(0, m_vbh);
            bgfx.setIndexBuffer(ibh);

            -- Set render states.
            bgfx.setState(state);

            -- Submit primitive for rendering to view 0.
            bgfx.submit(0, m_program);
        end
    end
    if(logFirst) then
        print("----- end 11*11 curbes -----\n")
    end
    logFirst = nil;
    -- Advance to next frame. Rendering thread will be kicked to
    -- process submitted rendering primitives.
    bgfx.frame();
    return 0;
end
local app_destroy = function ()
    print("app_destroy");
    -- [1~5]
    for i = 1, #m_ibh  do
        bgfx.destroy(m_ibh[i])
        bgfx.destroy(m_vbh);
        bgfx.destroy(m_program);
    end
end

app = bgfx.newApp(app_pre_init, app_init, app_draw, app_destroy);
app.start(app);